src/EFCore.PG/Storage/Internal/Mapping/NpgsqlListTypeMapping.cs

```diff
@@ -32,53 +35,60 @@
 namespace Npgsql.EntityFrameworkCore.PostgreSQL.Storage.Internal.Mapping
 {
     /// <summary>
-    /// Maps PostgreSQL arrays to .NET List{T}.
+    /// Maps PostgreSQL arrays to <see cref="List{T}"/>.
     /// </summary>
     public class NpgsqlListTypeMapping : RelationalTypeMapping
     {
+        /// <summary>
+        /// The CLR type of the list items.
+        /// </summary>
         public RelationalTypeMapping ElementMapping { get; }
 
         /// <summary>
-        /// Creates the default array mapping (i.e. for the single-dimensional CLR array type)
+        /// Creates the default list mapping.
         /// </summary>
         public NpgsqlListTypeMapping(RelationalTypeMapping elementMapping, Type listType)
-            : this(elementMapping.StoreType + "[]", elementMapping, listType)
-        {}
+            : this(elementMapping.StoreType + "[]", elementMapping, listType) {}
 
+        /// <inheritdoc />
         NpgsqlListTypeMapping(string storeType, RelationalTypeMapping elementMapping, Type listType)
-            : base(new RelationalTypeMappingParameters(
-                new CoreTypeMappingParameters(listType, null, CreateComparer(elementMapping, listType)), storeType
-            ))
-        {
-            ElementMapping = elementMapping;
-        }
+            : base(
+                new RelationalTypeMappingParameters(
+                    new CoreTypeMappingParameters(listType, null, CreateComparer(elementMapping, listType)), storeType))
+            => ElementMapping = elementMapping;
 
+        /// <inheritdoc />
         protected NpgsqlListTypeMapping(RelationalTypeMappingParameters parameters, RelationalTypeMapping elementMapping)
-            : base(parameters) {}
+            : base(parameters)
+            => ElementMapping = elementMapping;
 
+        /// <inheritdoc />
         public override RelationalTypeMapping Clone(string storeType, int? size)
             => new NpgsqlListTypeMapping(StoreType, ElementMapping, ClrType);
 
+        /// <inheritdoc />
         public override CoreTypeMapping Clone(ValueConverter converter)
             => new NpgsqlListTypeMapping(Parameters.WithComposedConverter(converter), ElementMapping);
 
+        /// <inheritdoc />
         protected override string GenerateNonNullSqlLiteral(object value)
         {
-            // TODO: Duplicated from NpgsqlArrayTypeMapping
-            var arr = (Array)value;
+            var list = (IList)value;
```

Comment:
@austindrenski I'm not too sure about this...

For one thing multidimensional arrays seem to implement `List` - but not `List<T>`. This is odd, and I suspect it to be .NET legacy behavior which was kept to maintain backwards compatibility.

At the very least, please add a specific check for multidimensional arrays, throwing `NotSupportedException` - it's an error to render a multidimensional .NET array as a one-dimensional PostgreSQL array...

(it would have been better to allow reviewing this non-cosmetic change before merging, but no big deal :)).

Reply:
Admittedly, I wasn't thinking about multidimensional arrays here...which underscores your point about reviews—that's my bad.

Since we don't support jagged arrays, I assumed that `List<T>` would only be mappable to one-dimensional PostgreSQL arrays. (Since `List<List<T>>` is more or less the equivalent of `T[][]`.) 

Did you have something else in mind?

Reply:
See some additional thoughts in https://github.com/npgsql/Npgsql.EntityFrameworkCore.PostgreSQL/pull/552#discussion_r205986013.

As I wrote there, we basically need to separate between multidimensional arrays and jagged arrays. The latter really aren't supported by PostgreSQL and a clear, informative message should say so. I agree that `List<List<T>` is just another form of a jagged array and should throw in the same way.

There's no such thing as a multidimensional array with `List<T>`, so I don't think you need to think about that here in `NpgsqlListTypeMapping` (as opposed to `NpgsqlArrayTypeMapping`).

Aside from that, I'm simply not sure there's value in support non-generic `List` - I'm assuming part of you change was to support that. It's probably healthier to restrict support to generic `List<T>` only.

Hope this is clear, we're discussing several different things here...

Reply:
So, I think we're on the same page. But I'm still a little lost on what you mean by the non-generic `List`: 

> Aside from that, I'm simply not sure there's value in support non-generic `List` - I'm assuming part of you change was to support that.

1. My understanding is that we will only support `List<T>` where `T` is not a collection type. 
2. I cast the incoming parameter to `IList` so that I could access `IList.Count` for the loop below.
   - I'm using `IList` since the type mapper is non-generic, so there's no type parameter in scope for `IList<T>`/`List<T>`.

Reply:
Yeah, I wasn't very clear on that.

First, I'd rather we had a clear up-front check that `GenerateNonNullSqlLiteral()` gets the right kind of input, i.e. a generic `List<T>` where `T` corresponds to `ElementMapping`. Right now we throw with imprecise messages (multidimensional arrays, or just that the given object isn't a generic type).

Second, as I wrote in the other comment, I'd rather we had construction-time checks for the correctness of the mapping object itself, i.e. throw if we attempt to construct a mapping that would correspond to a jagged array (i.e. `T[][]`, `List<List<T>>`).

After all these checks it seems fine to cast to `IList` as we've verified that the types are correct.

Reply:
Got it. I'll put something together in the next few days.